### ISSUE TITLE
Feature - save family tree view settings

### ISSUE THREAD

**Issue by OJ7** [gramps-project/Gramps.js#67]
Whenever I go to the family tree view, I usually change it to vertical view with 5 ancestors/descendants. I have to do this each time I open up the website. It would be nice to have this persist across sessions.

**Comment by DavidMStraub** [comment]
Yes, good idea.

If somebody wants to implement this:

`GrampsjsViewGraph` has a property `settings` that is an object that contains all the settings (synced with LocalStorage). So, just need to implement two things:

- set view parameters based on a new settings key, if present
- update this settings key when a parameter is changed, using `updateSettings({someKey: someValue})` from `api.js`

**Comment by BlitzundBogen** [comment]
I could try to solve this. 😄

**Comment by DavidMStraub** [comment]
April 2024 update on the implementation: use `this.appState.settings` and `this.appState.updateSettings` (which are available in all components).

**Comment by kinslayer1982** [comment]
I guess adding the option to change the default values is related to this request. Or should it be a separate one?
For now I changed the default values in the .js file and mount it inside the container as a quick workaround.

**Comment by DavidMStraub** [comment]
The default values are the default values, then there are user settings which could be made persistent.

**Comment by kinslayer1982** [comment]
Well, that's a VERY simple point of view, don't you think? I bet you wouldn't like all software having everything hardcoded without the ability to set reasonable default values. According to your answer nothing like environment variables would exist, no way of customizing the deployment of software, containers, vms...

Let's take the invoicing software I set up for my wife: The default currency of the software is $. This doesn't mean that she has to use $ everywhere. No, I can configure the system to use € as default currency for every invoice that gets created. Which again does not prevent her to set the currency for specific customers to $, €, or whatever...

Yes, you need defaults so the software may run even if nothing is specified at the time of deployment. But it doesn't mean that these defaults are suitable for every situation / use-case.

I'm just offering ideas to you and others that are not comfortable with the hardcoded defaults and no way to change that.
If you don't want that I can stop making suggestions or try and investigate issues I and share what I find.

**Comment by DavidMStraub** [comment]
This issue is about saving preferences for the family tree view. It's completely unnecessary to start an argument about the necessity of preferences. If the user sets their view to use € instead of $, that's a preference, and the default (which is used when the user hasn't set a preference) is one of the two.

**Comment by kinslayer1982** [comment]
> This issue is about saving preferences for the family tree view. It's completely unnecessary to start an argument about the necessity of preferences. If the user sets their view to use € instead of $, that's a preference, and the default (which is used when the user hasn't set a preference) is one of the two.

Yes and no. You say only two things exist: the default value and the users preference. I say there can (and in most cases are) more: A hardcoded default value (which allows a software to run when nothing else is specified), a "default" value that overrides the hardcoded one set for example by an admin at deployment of the software, and after that the users preferences.

My question was simply if having the ability to set a different default value (like using environment variables in docker, or by other means) related closely enough to making user preferences work to combine these two points, or if this should be a separate issue/feature request.

You complain about duplicate issue reports. (Ok, I absolutely understand why. Sorry if I didn't see it as a same issue because of wording as english is not my native language.)
Asking if it should be a separate issue/request you give an answer that TO ME gives the impression you don't care. 
Plus closing issues marking them as completed when they are not.

My intent was not to argue. But I've had my share of developers who only complain, no matter what or how you say something.
Sorry if I got you wrong, but your answer didn't make sense to me.

**Comment by DavidMStraub** [comment]
> You complain about duplicate issue reports. 

No I don't. I wrote:

> Agreed, but that's a duplicate of https://github.com/gramps-project/gramps-web/issues/67.

Is that a complaint?

> Asking if it should be a separate issue/request you give an answer that TO ME gives the impression you don't care.

In my workplace, we had "always assume good intent" as one of our core values. Think about it. Look at this project, about the effort I put in, and then think twice.

>  Sorry if I got you wrong, but your answer didn't make sense to me.

Fair, let me break it down for you again.

This is an issue in the Gramps Web frontend repo. Gramps Web uses a REST API as backend. The REST API is pretty much agnostic of the frontend, so we can't (or let's say won't) put frontend specific settings somewhere in the backend.

The frontend has default values (whether the current ones are good or not is up for discussion). Then it *could* have settings that persist in localStorage.

That was the long version of my short comment above.

By the way: since apparently you also have experience in maintaining software, you probably know that all of this takes a lot of time. So often, comments are brief to save time. Better a brief answer in a short amount of time.

What your are asking is essentially if you could modify *frontend default settings* as well. We only have that in a few exceptional cases via `grampsjsConfig` (oops the [docs](https://www.grampsweb.org/install_setup/frontend-config/) are outdated), but I don't think family tree view settings should go in there. So, in short, there is no need to open a separe issue.

**Comment by kinslayer1982** [comment]
> > You complain about duplicate issue reports.
> 
> No I don't. I wrote:
> 
> > Agreed, but that's a duplicate of [#67](https://github.com/gramps-project/gramps-web/issues/67).
> 
> Is that a complaint?

I didn't mean a specific one. Looking through issues I got the impression now and then. But complain maybe is too hard.
 
> > Asking if it should be a separate issue/request you give an answer that TO ME gives the impression you don't care.
> 
> In my workplace, we had "always assume good intent" as one of our core values. Think about it. Look at this project, about the effort I put in, and then think twice.

Well, like I said in my personal experience shows something else.

> > Sorry if I got you wrong, but your answer didn't make sense to me.
> 
> Fair, let me break it down for you again.
> 
> This is an issue in the Gramps Web frontend repo. Gramps Web uses a REST API as backend. The REST API is pretty much agnostic of the frontend, so we can't (or let's say won't) put frontend specific settings somewhere in the backend.
> 
> The frontend has default values (whether the current ones are good or not is up for discussion). Then it _could_ have settings that persist in localStorage.
> 
> That was the long version of my short comment above.
> 
> By the way: since apparently you also have experience in maintaining software, you probably know that all of this takes a lot of time. So often, comments are brief to save time. Better a brief answer in a short amount of time.

It wasn't about the lenght. Something like "Not going to make default values adjustable for now" is even shorter and wouldn't have given me reason to further talk about it, if you know what I mean.

> What your are asking is essentially if you could modify _frontend default settings_ as well. We only have that in a few exceptional cases via `grampsjsConfig` (oops the [docs](https://www.grampsweb.org/install_setup/frontend-config/) are outdated), but I don't think family tree view settings should go in there. So, in short, there is no need to open a separe issue.

Ok.

Again, I didn't mean any offense. Just a misunderstanding.

**Comment by DavidMStraub** [comment]
Fixed by #911!
